### PR TITLE
[4/N][CP] Support interleave strategy for cp v2

### PR DESCRIPTION
**Pull request by hzh0425**
## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :hourglass_flowing_sand: [Run #30524785696](https://github.com/sgl-project/sglang/actions/runs/30524785696)
Latest PR Test (Extra): :hourglass_flowing_sand: [Run #30524785202](https://github.com/sgl-project/sglang/actions/runs/30524785202)

### CODE REVIEW

**Comment by hzh0425** [comment]
/rerun-test test/registered/cp/test_deepseek_v32_cp_single_node.py

**Comment by hzh0425** [comment]
/rerun-test test/registered/cp/test_deepseek_v32_cp_single_node.py

**Comment by hzh0425** [comment]
/rerun-test test/registered/cp/test_deepseek_v32_cp_single_node.py

**Comment by Fridge003** [comment]
We might need to rebase after this PR merges https://github.com/sgl-project/sglang/pull/30992

File: test/registered/cp/test_deepseek_v32_cp_single_node.py [hunk 1]
```diff
@@ -152,5 +152,71 @@ def test_a_gsm8k(
             self.assertGreater(metrics["score"], 0.935)
 
 
+class TestDeepseekV32CPV2Interleave(CustomTestCase):
```
Comment (Fridge003) [inline test/registered/cp/test_deepseek_v32_cp_single_node.py]: Let's create a new file `test_dsa_prefill_cp.py` for CP_V2 tests.
Instead of dsv32, we can use glm52 by  deleting `test/registered/cp/test_glm52_cp_index_share.py` and moving that test to the newly created file.

`test/registered/cp/test_deepseek_v32_cp_single_node.py` can be renamed as `test_dsa_prefill_cp_legacy.py`, where `SGLANG_ENABLE_CP_V2` is set to 0

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -5033,9 +5033,15 @@ def _handle_legacy_cp_arguments(self):
...
 
         mode = strategy_to_legacy_mode[self.cp_strategy]
-        use_dsa_legacy_aliases = self.enable_dsa_prefill_context_parallel or getattr(
-            self, "attention_backend", None
-        ) in ("dsa", "dsv4")
+        attention_backend = getattr(self, "attention_backend", None)
+        if attention_backend is None:
+            from sglang.srt.arg_groups.overrides import resolved_view
+
+            attention_backend = getattr(resolved_view(self), "attention_backend", None)
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py]: This part should have been fixed in latest main

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py [hunk 3]
```diff
@@ -557,10 +558,22 @@ def forward_absorb_prepare(
...
         if (dsa_prefill_cp or mla_prefill_cp) and not defer_kv_gather_until_after_rope:
-            # support allgather+rerrange
-            k_nope, k_pe = self.rebuild_cp_kv_cache(
-                latent_cache, forward_batch, k_nope, k_pe
-            )
+            if is_cp_v2_active(forward_batch):
+                # CP-v2: AllGather latent KV via strategy.materialize_full_kv
+                from sglang.srt.layers.cp.base import get_cp_strategy
+
+                k_nope, k_pe = get_cp_strategy().materialize_full_kv(
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py]: Can we put materialize_full_kv in `dsa_backend.py`. Any CP related logics should be decoupled with modeling files

File: python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py [hunk 4]
```diff
@@ -557,10 +558,22 @@ def forward_absorb_prepare(
...
+                    forward_batch,
```
Comment (Fridge003) [inline python/sglang/srt/models/deepseek_common/attention_forward_methods/forward_mla.py]: Can we modify the changes in https://github.com/sgl-project/sglang/pull/29161, so that it's supported by CP_V2, otherwise CP_V2 will break on trtllm attention backend + Blackwell hardware

For example, the `_all_gather_dsa_trtllm_fp8_kv` function in #29161 should be a member function of CP strategies

File: python/sglang/srt/server_args.py [hunk 5]
```diff
@@ -5033,9 +5033,15 @@ def _handle_legacy_cp_arguments(self):
             return
```
Comment (Fridge003) [inline python/sglang/srt/server_args.py:6176]: There is a bunch of util functions for dsa cp in `python/sglang/srt/layers/attention/dsa/utils.py`.

1. We need to rewrite some of them, so they are aligned with usage of CP_V2
For example, `is_dsa_enable_prefill_cp` should be `get_parallel().attn_cp_size > 1 and is_dsa_model` instead, where is_dsa_model can be something like `is_deepseek_dsa(model_hf_config)`

Image

2. There are some functions for splitting or gathering data, like `dsa_cp_round_robin_split_q_seqs`. We need to make them a method of interleave strategy, then delete them from`dsa/utils.py`

File: python/sglang/srt/layers/cp/interleave.py [hunk 6]
```diff
@@ -94,14 +201,39 @@ def run_attention(
         device: Any,
         attn_fn,
         attention_backend: CPAttentionBackendKind = CPAttentionBackendKind.FLASH_ATTENTION,
+        **kwargs,
     ) -> Any:
-        raise NotImplementedError(
-            "Interleave attention dispatch will land in a follow-up PR"
-        )
```
Comment (Fridge003) [inline python/sglang/srt/layers/cp/interleave.py:100]: Add a comment on why here is `pass`

File: python/sglang/srt/layers/cp/utils.py [hunk 7]
```diff
@@ -35,6 +36,7 @@
 
 CP_V2_DEFAULT_MODEL_CLASSES = frozenset(
     {
+        "DeepseekV32ForCausalLM",
```
Comment (Fridge003) [inline python/sglang/srt/layers/cp/utils.py:44]: Maybe include `GlmMoeDsaForCausalLM`

File: python/sglang/srt/model_executor/runner/eager_runner.py [hunk 8]
```diff
@@ -346,14 +342,23 @@ def _execute_extend(
...
                     )
             elif cp_v2_active:
-                # CP-V2: drive .model directly to gather across CP ranks before logits.
-                hidden_states = model_runner.model.model(
-                    forward_batch.input_ids,
-                    forward_positions,
+                # CP-V2: shard inputs at the boundary, drive .model directly to
+                # gather across CP ranks before logits. The context manager keeps
+                # the model CP-agnostic by sharding/restoring spec_info in-place.
+                with cp_shard_model_inputs(
```
Comment (Fridge003) [inline python/sglang/srt/model_executor/runner/eager_runner.py]: Why are we creating this decorator for sharding inputs?

File: python/sglang/srt/layers/cp/interleave.py [hunk 9]
```diff
@@ -49,43 +59,140 @@ class InterleaveCPStrategy(ContextParallelStrategy):
...
+        return self._round_robin_shard(x)
 
     def shard_position_ids(self, positions: Any, forward_batch) -> Any:
-        raise NotImplementedError(
-            "Interleave position-id sharding will land in a follow-up PR"
-        )
+        return self._round_robin_shard(positions)
+
+    def _round_robin_shard(self, input_: Any) -> Any:
```
Comment (Fridge003) [inline python/sglang/srt/layers/cp/interleave.py]: Rename to _interleave_shard

File: python/sglang/srt/layers/attention/dsa/dsa_indexer.py [hunk 10]
```diff
@@ -618,6 +620,10 @@ def _get_q_k_bf16(
             self.alt_stream.wait_stream(current_stream)
             query = self._maybe_rotate(query)
 
+            if is_cp_v2_active(forward_batch):
+                current_stream.wait_stream(self.alt_stream)
+                return query, key, weights_raw
+
             with torch.cuda.stream(self.alt_stream):
```
Comment (Fridge003) [inline python/sglang/srt/layers/attention/dsa/dsa_indexer.py:637]: Instead of doing `get_cp_strategy().run_indexer`, I think we should do 
```
if is_cp_v2_active(forward_batch):
    key = get_cp_strategy().materialize_full_indexer_k_cache() 
```
at this place.

Comment (hzh0425) [reply python/sglang/srt/layers/attention/dsa/dsa_indexer.py:637]: done

File: python/sglang/srt/layers/attention/dsa/utils.py [hunk 11]
```diff
@@ -200,6 +210,15 @@ def pad_dsa_cache_seqlens(forward_batch: "ForwardBatch", dsa_cache_seqlens):
 
 
 def can_dsa_cp_split(seq_len: int, cp_size: int, use_dsa: bool, forward_batch):
+    if (
```
Comment (hzh0425) [inline python/sglang/srt/layers/attention/dsa/utils.py:219]: Check in advance to prevent short sequences from entering cp.

File: python/sglang/srt/layers/attention/dsa_backend.py [hunk 12]
```diff
@@ -923,12 +942,19 @@ def init_forward_metadata(self, forward_batch: ForwardBatch):
             )
 
             if can_dsa_prefill_cp_round_robin_split(forward_batch):
-                seqlens_expanded = dsa_cp_round_robin_split_data(seqlens_expanded)
-                extend_seq_lens_cpu, extend_seq_lens, bs_idx_cpu, bs_idx = (
-                    dsa_cp_round_robin_split_q_seqs(
-                        extend_seq_lens_cpu, extend_seq_lens
+                if is_cp_v2_active(forward_batch):
+                    strategy = get_cp_strategy()
+                    seqlens_expanded = strategy.shard_local_tokens(seqlens_expanded)
```
Comment (hzh0425) [inline python/sglang/srt/layers/attention/dsa_backend.py:974]: move some v1 logics to v2 strategy here

File: python/sglang/srt/layers/cp/interleave.py (hunk 6 above)
Comment (hzh0425) [reply python/sglang/srt/layers/cp/interleave.py:100]: done

File: python/sglang/srt/layers/cp/utils.py (hunk 7 above)
Comment (hzh0425) [reply python/sglang/srt/layers/cp/utils.py:44]: done

File: python/sglang/srt/server_args.py (hunk 5 above)
Comment (hzh0425) [reply python/sglang/srt/server_args.py:6176]: done

**Comment by hzh0425** [comment]
cc @Fridge003 I've adjusted a version based on your review. ～

**Comment by hzh0425** [comment]
/rerun-test registered/cp/test_glm52_cp_index_share.py registered/cp/test_gpt_oss_4gpu_mxfp4_cp.py

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

**Review by Fridge003 — APPROVED** [review]

**Comment by Fridge003** [comment]
/rerun-test test/registered/models_e2e/test_dsa_glm52_cache_layer_split.py

**Comment by hzh0425** [comment]
/rerun-test test/registered/cp/test_glm52_cp_index_share.py  registered/models_e2e/test_dsa_glm52_cache_layer_split.py test/registered/cp/test_dsa_prefill_cp.py
